lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -603,6 +603,120 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
   return tryToConcatenateStrings(AI, Builder);
 }
 
+/// This routine replaces the old witness method inst with a new one.
+void SILCombiner::replaceWitnessMethodInst(
+    FullApplySite Apply, WitnessMethodInst *WMI, SILBuilderContext &BuilderCtx,
+    const CanType &ConcreteType, const ProtocolConformanceRef ConformanceRef) {
+  SILBuilderWithScope WMIBuilder(WMI, BuilderCtx);
+  auto *NewWMI = WMIBuilder.createWitnessMethod(
+      WMI->getLoc(), ConcreteType, ConformanceRef, WMI->getMember(),
+      WMI->getType());
+  MutableArrayRef<Operand> Operands = Apply.getInstruction()->getAllOperands();
+  for (auto &Op : Operands) {
+    if (Op.get() == WMI)
+      Op.set(NewWMI);
+  }
+  if (WMI->use_empty())
+    eraseInstFromFunction(*WMI);
+}
+
+//  This function propagates concrete type of existential self argument using
+//  ProtocolConformanceAnalysis. The concrete type of self can be a class,
+//  struct, or an enum.
+//  For example, if C is the only class conforming to a protocol P, then
+//  transfrom from the following
+//  %4 = open_existential_ref %3 : $P to $@opened P
+//  %5 = witness_method $@opened P, #P.foo!1 : <Self
+//     where Self : P> (Self) -> () -> Int, %4 :  $@opened P :
+//     $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@guaranteed
+//     τ_0_0) -> Int
+//  %6 = apply %5<@opened P>(%4) : $@convention(witness_method:
+//     P) <τ_0_0 where τ_0_0 : P> (@guaranteed τ_0_0) -> Int
+//  TO
+//  %4 = open_existential_ref %3 : $P to $@opened P
+//  %7 = unchecked_ref_cast %4 : $@opened P to $C
+//  %8 = class_method %7 : $C, #C.foo!1 : (C) -> () -> Int,
```

Comment:
The actual replacement of the witness_method with class_method is done in the devirtualization pass, not this function. Also the SIL is just kind of hard to read. Can you instead just write a comment that says it replaces the witness_method instruction with one that has a concrete type, allowing other optimizations to devirtualize it later?